src/msg/async/AsyncConnection.h

```diff
@@ -124,9 +127,53 @@ class AsyncConnection : public Connection {
     return !out_q.empty();
   }
 
+   /**
+   * The DelayedDelivery is for injecting delays into Message delivery off
+   * the socket. It is only enabled if delays are requested, and if they
+   * are then it pulls Messages off the DelayQueue and puts them into the
+   * AsyncMessenger event queue.
+   * This is a nearly direct copy & paste from SimpleMessenger, and as
+   * such, there was a problem during AsyncConnection shutdown, fixed by
+   * checking whether delay_thread isn't already stopped and refraining
+   * from stopping it again.
+   */
+  class DelayedDelivery : public EventCallback {
+    set<uint64_t> register_time_events; // need to delete it if stop
+    std::deque<std::pair<utime_t, Message*> > delay_queue;
+    Mutex delay_lock;
+    AsyncMessenger *msgr;
+    EventCenter *center;
+
+   public:
+    explicit DelayedDelivery(AsyncMessenger *omsgr, EventCenter *c)
+      : delay_lock("AsyncConnection::DelayedDelivery::delay_lock"),
+        msgr(omsgr), center(c) { }
+    ~DelayedDelivery() {
+      for (auto &&i : register_time_events)
```

Comment:
`i` is still a universal reference not a rvalue reference. `auto` is better here. since `i` is just a `uint64_t`.